### PR TITLE
Add stac-arrow

### PR DESCRIPTION
**Pull request by gadomski**
## Description

Read and write **stac-geoparquet**.

## Checklist

- [x] Unit tests
- [x] Documentation, including doctests
- [x] Git history is linear
- [x] Commit messages are descriptive
- [x] (optional) Git commit messages follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] Code is formatted (`cargo fmt`)
- [x] `cargo test`
- [x] Changes are added to the CHANGELOG

### CODE REVIEW

**Comment by kylebarron** [comment]
> it looks like **arrow-json** is moving away from providing direct **serde_json** support (e.g. this deprecated function: [docs.rs/arrow-json/51.0.0/arrow_json/writer/fn.record_batches_to_json_rows.html](https://docs.rs/arrow-json/51.0.0/arrow_json/writer/fn.record_batches_to_json_rows.html)). 

Indeed. See https://github.com/apache/arrow-rs/pull/5318. I believe serde-json support will be removed in a future arrow-rs release.

**Comment by gadomski** [comment]
I did a quick-and-dirty benchmark ([stac-arrow/benches/read.rs](https://github.com/stac-utils/stac-rs/pull/256/files#diff-5323c5b3b6fc55ccbaff24702666be29c423b07ecf047dc42158439e4c877374)) and it looks like it's worth it to use/port the deprecated `record_batches_to_json_rows` for the "need to go through serde_json" case:

```text
read/record_batches_to_json_rows    time:   [170.61 µs 174.72 µs 178.93 µs]
read/writer                         time:   [224.61 µs 231.71 µs 241.36 µs]
```

**Comment by kylebarron** [comment]
If there's anything in geoarrow-rs that you'd like to use that I can help with, let me know.

**Comment by gadomski** [comment]
@kylebarron I think I've got a working start. There's obviously some/many edge cases in the geoarrow/geoparquet writing that I'll need to handle later, but for now [stac-geoparquet](https://github.com/stac-utils/stac-geoparquet) can read what I write so it's a start.

Some simple benchmarking against 10k sentinel-2 items indicates we might get some performance gains:

```shell
$ hyperfine --warmup 3 'target/release/stac convert ~/Desktop/longmont.json ~/Desktop/longmont-rs.parquet' 'python ~/Desktop/to_stac_geoparquet.py'
Benchmark 1: target/release/stac convert ~/Desktop/longmont.json ~/Desktop/longmont-rs.parquet
  Time (mean ± σ):      1.322 s ±  0.051 s    [User: 0.971 s, System: 0.246 s]
  Range (min … max):    1.265 s …  1.410 s    10 runs
 
Benchmark 2: python ~/Desktop/to_stac_geoparquet.py
  Time (mean ± σ):      2.374 s ±  0.061 s    [User: 2.111 s, System: 0.542 s]
  Range (min … max):    2.317 s …  2.490 s    10 runs
 
Summary
  target/release/stac convert ~/Desktop/longmont.json ~/Desktop/longmont-rs.parquet ran
    1.80 ± 0.08 times faster than python ~/Desktop/to_stac_geoparquet.py
```

Query:

```shell
target/release/stac search https://earth-search.aws.element84.com/v1 \                                                                           
    -c sentinel-2-c1-l2a \
    --max-items 10000 \
    --sortby='-properties.datetime' \
    --intersects '{"type":"Point","coordinates":[-105.1019,40.1672]}' > ~/Desktop/longmont.json
```
 
Benchmark script:

```python
import json

import stac_geoparquet.arrow
from pyarrow import Table

with open("/Users/gadomski/Desktop/longmont.json") as f:
    items = json.load(f)["features"]
table = Table.from_batches(stac_geoparquet.arrow.parse_stac_items_to_arrow(items))
stac_geoparquet.arrow.to_parquet(table, "/Users/gadomski/Desktop/longmont-py.parquet")
```

File: scripts/requirements.in [hunk 1]
```diff
@@ -1,2 +1,2 @@
 stac-api-validator
-stac-geoparquet
+stac-geoparquet<0.5  # v0.5 uses geoparquet 1.1, but geoarrow-rs is geoparquet 1.0
```
Comment (kylebarron) [inline scripts/requirements.in:2]: We should probably have an option in stac-geoparquet for whether to write 1.0 or 1.1 compliant metadata.

🥲 I have some support in geoarrow-rs for geoparquet 1.1 but not full support yet.

**Comment by kylebarron** [comment]
> Some simple benchmarking against 10k sentinel-2 items indicates we might get some performance gains (though I'm not thinking about compression yet so could be a bad comparison):

I'm a little surprised the Rust version isn't more than 1.8x faster. I suppose on the Python side once the JSON is converted to Arrow, it's fully compiled code anyways.

**Comment by kylebarron** [comment]
Ah, also this is the most efficient way to convert JSON to Arrow (though it uses the most memory) because it can pass those dicts directly to pyarrow, which then presumably is fully compiled in its conversion to Arrow

> ```
> table = Table.from_batches(stac_geoparquet.arrow.parse_stac_items_to_arrow(items))
> ```

**Comment by kylebarron** [comment]
The goal for geoarrow-rs' python bindings for 0.3 is for it to exclusively use `arro3.core` objects, and not define _any_ new classes itself. So `GeoTable` will be replaced by returning an `arro3.core.table`

**Comment by gadomski** [comment]
Marking this as draft until **geoarrow-rs** goes v0.3.

**Comment by kylebarron** [comment]
> Marking this as draft until **geoarrow-rs** goes v0.3.

sounds like you've seen some breaking changes 🫣

**Comment by kylebarron** [comment]
I've also been making a decent number of changes on the Rust side, like 3D support

**Comment by gadomski** [comment]
Superseded by #287

**Pull request closed without merging**